src/other/utils.jl

```diff
@@ -83,3 +83,37 @@ else
 end
 
 funname(c::ComposedFunction) = Symbol(funname(c.outer), :_, funname(c.inner))
+
+"""
+    tforeach(f, x::AbstractArray; basesize::Integer)
+
+Apply function `f` to each entry in `x` in parallel, spawning
+one separate task for each block of `basesize` entries.
+
+A number of task higher than `Threads.nthreads()` may be spawned,
+since that can allow for a more efficient load balancing in case
+some threads are busy (nested parallelism).
+"""
+function tforeach(f, x::AbstractArray; basesize::Integer)
+    @static if VERSION >= v"1.4"
+        nt = Threads.nthreads()
+        if nt > 1 && length(x) > basesize
+            @sync for p in partition(x, basesize)
```

Comment:
is `partition(x, basesize)` really optimal?

I ask because I have noticed:
```
julia> collect(DataFrames.partition(1:21, 10))
3-element Vector{UnitRange{Int64}}:
 1:10
 11:20
 21:21
```
which means that we would have one small dangling iteration while it would be probably better to consume it (but maybe it is OK?).

Reply:
Very good point. I've pushed a commit which increases the size to ensure all chunks have at least `basesize` entries. In your example, this would give only two chunks, which is more in line with the docstring. Please check that it's correct as I always get these things wrong. :-)

Reply:
BTW, the current approach is quite good when the number of tasks is large compared with the number of threads, as work will be distributed quite evenly over threads. But it's not completely optimal if there are few tasks. For example, with 3M rows and 2 threads, 3 tasks will be started with each 1M rows, so one thread will finish much before the other one. It would be more efficient to start two tasks with 1.5M rows each.

But there are more complex cases: imagine we have 4M rows and 3 threads, but one thread is actually busy doing something else. If we start 3 tasks instead of 4, we end up in the situation described in the previous paragraph. Not sure how to fix that in general. Maybe we shouldn't care as the 1M threshold is chosen precisely because it takes very little time to process a chunk of that size. If it's a problem that one thread gets one more chunk than other threads, maybe we should simply decrease the threshold...

Reply:
This fix is incorrect:
```
julia> len = 122
122

julia> basesize = 10
10

julia> basesize′ = cld(len, fld(len, basesize))
11

julia> collect(Iterators.partition(1:len, basesize′))
12-element Vector{UnitRange{Int64}}:
 1:11
 12:22
 23:33
 34:44
 45:55
 56:66
 67:77
 78:88
 89:99
 100:110
 111:121
 122:122
```
The way to fix it is something like:
```
function getranges(len, batchsize)
    partitions = div(len, batchsize)
    d , r = divrem(len, partitions)
    [((i - 1) * d + min(r, i - 1) + 1) .+ (0:d - 1 + (i ≤ r ? 1 : 0)) for i in 1:partitions]
end
```
but it is complex so I am not sure if it is worth the effort?

Reply:
> If it's a problem that one thread gets one more chunk than other threads, maybe we should simply decrease the threshold...

That is the point I think (and also why I am considering my fix overly complex) - I assume that 1M rows is small enough to be negligible.

Reply:
I think it's worth it. But can't we find a simplest solution (i.e. one I can understand :-p)? At #2491 I had copied a solution I found on Discourse. Do you think it's correct?
```julia
julia> len = 122
122

julia> basesize = 10
10

julia> npartitions = div(len, basesize)
12

julia> collect((1 + ((i - 1) * len) ÷ npartitions):((i * len) ÷ npartitions) for i in 1:npartitions)
12-element Vector{UnitRange{Int64}}:
 1:10
 11:20
 21:30
 31:40
 41:50
 51:61
 62:71
 72:81
 82:91
 92:101
 102:111
 112:122
```

Reply:
The https://github.com/JuliaData/DataFrames.jl/pull/2491/files#diff-b452a777f1ba151f78b9d66f93f4bdb7a8148c5cbc5329fcbcea56954fe7c7a5R203 implementation is OK, but it only spawns as many tasks as you have threads so using `÷` (which is expensive) is not a problem.

My implementation has the same approach as the one you reference to but avoids `÷`. They essentially do the following:
1. calculate the required number of partitions (rounding down so that we are sure each group will get enough elements)
2. then each group gets either `k` or `k+1` elements (as number of groups might not be a divisor of length)
3. the complexity in the codes is to ensure that this `+1` is added as many times as the reminder from the division

However, both your code and my code are incorrect as they fail if `len < basesize`. Here is your corrected code along with correctness tests:

```
function fun(len::Int64, basesize) # Int64 is relevant here to avoid overflow on 32-bit machines
    npartitions = max(1, div(len, basesize))
    x = collect((1 + ((i - 1) * len) ÷ npartitions):((i * len) ÷ npartitions) for i in 1:npartitions)

    # correctness tests
    @assert length(x) == npartitions
    @assert reduce(vcat, x) == 1:len
    vmin, vmax = extrema(length, x)
    @assert vmin + 1 == vmax || vmin == vmax
    @assert len < basesize || vmin >= basesize

    return x
end

for len in 1:1000, basesize in 1:100
    fun(len, basesize)
end
```

Reply:
Thanks. I've pushed a commit using your implementation.

FWIW I've realized that `Threads.@threads` also has similar code: https://github.com/JuliaLang/julia/blob/c19d389a00b4a15058c16ac46a991aa312c6fc40/base/threadingconstructs.jl#L47-L83